indev/libinput.c

```diff
@@ -410,39 +493,55 @@ static void reset_scanned_devices(void) {
 static void read_pointer(libinput_drv_state_t *state, struct libinput_event *event) {
   struct libinput_event_touch *touch_event = NULL;
   struct libinput_event_pointer *pointer_event = NULL;
+  libinput_lv_event_t *evt = NULL;
   enum libinput_event_type type = libinput_event_get_type(event);
 
+  /* We only care about these events */
+  if (type != LIBINPUT_EVENT_TOUCH_MOTION &&
+      type != LIBINPUT_EVENT_TOUCH_DOWN &&
+      type != LIBINPUT_EVENT_TOUCH_UP &&
+      type != LIBINPUT_EVENT_POINTER_MOTION &&
+      type != LIBINPUT_EVENT_POINTER_BUTTON)
+    return;
+
   /* We need to read unrotated display dimensions directly from the driver because libinput won't account
    * for any rotation inside of LVGL */
   lv_disp_drv_t *drv = lv_disp_get_default()->driver;
 
+  /* ignore more than 2 fingers as it will only confuse LVGL */
+  touch_event = libinput_event_get_touch_event(event);
+  if (touch_event && libinput_event_touch_get_slot(touch_event) > 1)
+    return;
+
+  evt = new_event(state);
+
   switch (type) {
     case LIBINPUT_EVENT_TOUCH_MOTION:
-    case LIBINPUT_EVENT_TOUCH_DOWN:
-      touch_event = libinput_event_get_touch_event(event);
+    case LIBINPUT_EVENT_TOUCH_DOWN: {
       lv_coord_t x = libinput_event_touch_get_x_transformed(touch_event, drv->physical_hor_res > 0 ? drv->physical_hor_res : drv->hor_res) - drv->offset_x;
       lv_coord_t y = libinput_event_touch_get_y_transformed(touch_event, drv->physical_ver_res > 0 ? drv->physical_ver_res : drv->ver_res) - drv->offset_y;
       if (x < 0 || x > drv->hor_res || y < 0 || y > drv->ver_res) {
         break; /* ignore touches that are out of bounds */
       }
-      state->most_recent_touch_point.x = x;
-      state->most_recent_touch_point.y = y;
-      state->button = LV_INDEV_STATE_PR;
+      evt->point.x = x;
+      evt->point.y = y;
+      evt->pressed = LV_INDEV_STATE_PR;
       break;
+    }
     case LIBINPUT_EVENT_TOUCH_UP:
-      state->button = LV_INDEV_STATE_REL;
+      evt->pressed = LV_INDEV_STATE_REL;
       break;
     case LIBINPUT_EVENT_POINTER_MOTION:
       pointer_event = libinput_event_get_pointer_event(event);
-      state->most_recent_touch_point.x += libinput_event_pointer_get_dx(pointer_event);
-      state->most_recent_touch_point.y += libinput_event_pointer_get_dy(pointer_event);
-      state->most_recent_touch_point.x = LV_CLAMP(0, state->most_recent_touch_point.x, drv->hor_res - 1);
-      state->most_recent_touch_point.y = LV_CLAMP(0, state->most_recent_touch_point.y, drv->ver_res - 1);
+      evt->point.x += libinput_event_pointer_get_dx(pointer_event);
+      evt->point.y += libinput_event_pointer_get_dy(pointer_event);
```

Comment:
Upon testing with a trackpad, it doesn't let me move the cursor from the top left corner of the screen and always jumps back. I think it's caused by this part here. We used to add the delta from `libinput_event_pointer_get_dx/y` to the most recently read coordinates before. But here `evt` is a blank slot that we've zero'ed out so we're always starting in the top left corner regardless of where the cursor moved before.

We could look back in the ring buffer and add the value of `point` from the previous event. However, we'd also have to account for overrunning the two pointers in the buffer again and I'm not sure what value we'd set for `point` in that case. What might be better instead is to store `dx` and `dy` here and then evaluate either `x` and `y` or `dx` and `dy` in `libinput_read_state` by making use of `last_event`. I think we'd also have to change that function to always update `last_event` though regardless of whether more events are pending or not.

Reply:
Proposal how to fix this: https://github.com/calebccff/lv_drivers/pull/3